consensus/tendermint/tendermint.go

```diff
@@ -0,0 +1,637 @@
+						54: 		StartRound(0)
+
+					Fetching the relevant proposal implies the sender of the proposal was the proposer for that
+					height and round. Also, since only the proposals with valid value are added to the message set, the
+					validity of the proposal can be skipped.
+
+					There is no need to check decision_p[h_p] = nil since it is implied that decision are made
+					sequentially, i.e. x, x+1, x+2... .
+
+				*/
+
+				if p.ID != nil {
+					var (
+						proposal   Proposal[V, H, A]
+						precommits []Precommit[H, A]
+						vals       []A
+					)
+
+					for _, prop := range proposalsForHR[t.validators.Proposer(p.Height, p.Round)] {
+						if (*prop.Value).Hash() == *p.ID {
+							proposal = prop
+						}
+					}
+
+					for addr, valPrecommits := range precommitsForHR {
+						for _, precommit := range valPrecommits {
+							if *precommit.ID == *p.ID {
+								precommits = append(precommits, precommit)
+								vals = append(vals, addr)
+							}
+						}
+					}
+					if t.validatorSetVotingPower(vals) >= q(t.validators.TotalVotingPower(p.Height)) {
+						t.blockchain.Commit(t.state.height, *proposal.Value, precommits)
+
+						t.messages.deleteOldMessages(t.state.height)
+						t.state.height++
+						t.startRound(0)
+
+						continue
+					}
+				}
+
+				/*
+					Check the upon condition on line 47:
+
+						47: upon 2f + 1 {PRECOMMIT, h_p, round_p, ∗} for the first time do
+						48: schedule OnTimeoutPrecommit(h_p , round_p) to be executed after timeoutPrecommit(round_p)
+				*/
+
+				vals := slices.Collect(maps.Keys(precommitsForHR))
+				if !t.state.line47Executed && t.validatorSetVotingPower(vals) >= q(t.validators.TotalVotingPower(p.Height)) {
+					t.scheduleTimeout(t.timeoutPrecommit(p.Round), precommit, p.Height, p.Round)
+					t.state.line47Executed = true
+				}
+
+			case tm := <-t.timeoutsCh:
+				fmt.Println("Inside timeout case", tm)
+				switch tm.s {
+				case propose:
+					t.OnTimeoutPropose(tm.h, tm.r)
+				case prevote:
+					t.OnTimeoutPrevote(tm.h, tm.r)
+				case precommit:
+					t.OnTimeoutPrecommit(tm.h, tm.r)
+				}
+
+				i := slices.Index(t.scheduledTms, tm)
+				t.scheduledTms = slices.Delete(t.scheduledTms, i, i+1)
+			}
+			i++
+		}
+	}()
+}
+
+func (t *Tendermint[V, H, A]) Stop() {
+	fmt.Println("Stopping Tendermint...")
+	close(t.quit)
+	for _, tm := range t.scheduledTms {
+		tm.Stop()
+	}
+	fmt.Println("Waiting fo the Tendermint loop to exit")
+	t.wg.Wait()
+}
+
+func (t *Tendermint[V, H, A]) startRound(r uint) {
+	fmt.Println("Inside start round", r)
+	if r != 0 && r <= t.state.round {
+		return
+	}
+
+	t.state.round = r
+	t.state.step = propose
+
+	t.state.line34Executed = false
+	t.state.line36Executed = false
+	t.state.line47Executed = false
+
+	if p := t.validators.Proposer(t.state.height, r); p == t.nodeAddr {
+		var proposalValue *V
+		if t.state.validValue != nil {
```

Comment:
Is state.validValue not always nil? I don't see where it's set

Reply:
Please see the latest changes, it is set when upon condition on line 36 becomes true.